nearuplib/localnet.py

```diff
@@ -0,0 +1,105 @@
+import sys
+import argparse
+import configparser
+from subprocess import Popen
+from shutil import rmtree
+from os import mkdir, kill, unlink
+from signal import SIGTERM
+from os.path import exists, expanduser, join
+import json
+
+
+def run_binary(path, home, action, shards=None, validators=None, non_validators=None, boot_nodes=None, output=None):
+    command = [path, '--home', home, action]
+    if shards:
+        command.extend(['--shards', str(shards)])
+    if validators:
+        command.extend(['--v', str(validators)])
+    if non_validators:
+        command.extend(['--n', str(non_validators)])
+    if boot_nodes:
+        command.extend(['--boot-nodes', boot_nodes])
+
+    if output:
+        stdout = open(f'{output}.out', 'w')
+        stderr = open(f'{output}.err', 'w')
+    else:
+        stdout = None
+        stderr = None
+
+    return Popen(command, stdout=stdout, stderr=stderr)
+
+
+def run(args):
+    if args.overwrite:
+        if exists(args.home):
+            print("Removing old data.")
+            rmtree(args.home)
+
+    if not exists(args.home):
+        run_binary(args.binary_path, args.home,
+                   'testnet', shards=args.num_shards, validators=args.num_nodes).wait()
+
+    # Edit args files
+    for i in range(0, args.num_nodes):
+        args_json = join(args.home, f'node{i}', 'config.json')
+        with open(args_json, 'r') as f:
+            data = json.load(f)
+        data['rpc']['addr'] = f'0.0.0.0:{3030 + i}'
+        data['network']['addr'] = f'0.0.0.0:{24567 + i}'
+        with open(args_json, 'w') as f:
+            json.dump(data, f, indent=2)
+
+    # Load public key from first node
+    with open(join(args.home, f'node0', 'node_key.json'), 'r') as f:
+        data = json.load(f)
+        pk = data['public_key']
+
+    # Recreate log folder
+    rmtree('logs', ignore_errors=True)
+    mkdir('logs')
+
+    # Spawn network
+    pid_fd = open('node.pid', 'w')
+    for i in range(0, args.num_nodes):
+        proc = run_binary(args.binary_path, join(args.home, f'node{i}'), 'run',
+                          boot_nodes=f'{pk}@127.0.0.1:24567' if i > 0 else None, output=f'logs/node{i}')
+        print(proc.pid, file=pid_fd)
+    pid_fd.close()
```

Comment:
This is clever! So if i understand correctly, nearup will exit after several run_binary. The nodes are still running in current shell? Until you localnet --stop it?

Reply:
Yes, that's correct